Smoke me/khw hvds pr

khwilliamson opened the pull request:
refactor regpiece

khwilliamson committed:
regcomp.c: Move test earlier in regpiece()

Prior to this commit, the code checked if this was a particular
quantifier; if so handled that, then checked if it were some other
quantifier, returning if not.

This commit changes to check first if it is any quantifier, using the
macro that checks for all of them, returning if not.

Thus after the new code we know it is a quantifier and this allows later
commits to simplify

khwilliamson committed:
regcomp.c: Combine common code

This bit of code occurs in two places.  Move it so that it gets executed
before the split, so it only has to be specified once.

khwilliamson committed:
regcomp.c: White-space only

Outdent since the previous commit removed a surrounding block

khwilliamson committed:
regcomp.c: Move some code in regpiece()

This moves some code, making it part of an if/else.  As a result, several
gotos are eliminated, and the code is in a more logical order.

khwilliamson committed:
regcomp.c: White-space only

Outdent since the previous commit removed a surrounding block.
And convert some tabs to blanks

khwilliamson committed:
regcomp.c: Rmv misleading comment

This NOTREACHED's presence just before an else would make one think
there is something funny going one, when there isn't.

khwilliamson committed:
regcomp.c: Move code to earlier in regpiece()

This is in preparation for a future commit that will change things so
this would need to be done earlier than the changes.

khwilliamson committed:
regcomp.c: Rmv #if 0 bitrotted code

This code has been #ifdef'd out for 20 years, and shows bit rot.  It was
kept around in case there was a problem with the replacement for it, but
that hasn't happened and is very unlikely to happen; in the meantime
this is a visual distraction to the code flow.

khwilliamson committed:
regcomp.c: regpiece() 0 times anything is 0

Prior to this commit, regpiece() gave width to many quantified values.
This is wrong.  If something is zero width, repeating it doesn't cause
it to gain width.  I don't know if this led to wrong pattern matching
results, but I imagine it led to slower results in some cases.

khwilliamson committed:
regcomp.c: Change label name; rmv extraneous goto

The name was misleading.  There are other things being done here.  And
previous restructuring led to a goto immediately prior to where it went
to.

khwilliamson committed:
regcomp.c: regpiece: Consolidate code

There is a common place these three occurrences can be placed at,

khwilliamson committed:
regcomp.c: regpiece(): Refactor two 'if's

I think this makes it clearer the commonalities of the * and +
quantifiers.

khwilliamson committed:
regcomp.c: regpiece(): More comments; white-space

hvds commented:
@khwilliamson It may take me a few days to get to this.

hvds commented on regcomp.c:
```diff
@@ -12730,6 +12730,21 @@ S_regpiece(pTHX_ RExC_state_t *pRExC_state, I32 *flagp, U32 depth)
+
+        */
+
+        if (!(flags&HASWIDTH) && op != '?')
```
Note that the `!(flags & HASWIDTH)`  is redundant now; and `op != '?'` could usefully be changed to `max == REG_INFTY` (with corresponding changes to the message), to cover also `(empty){1,}`.

khwilliamson replied:
I didn't touch anything within in #if 0 block, as it would be extra work to try to figure out what was meant, and who knows how much bit rot it had already been subject to.  Realistically, this has been this way for  20 years, lst touched  by Jeffrey Friedl, I very much doubt that tt will get reinstated, and I don't think it's worth trying to figure out what it should do in the nlikely event that we would need to inclde it.  So, I favor leaving it the way it is, or simply removing the whole block

hvds replied:
I feel we should improve it or remove it, so let's remove it.

khwilliamson replied:
will do

hvds commented on regcomp.c:
```diff
@@ -12745,6 +12745,16 @@ S_regpiece(pTHX_ RExC_state_t *pRExC_state, I32 *flagp, U32 depth)
           vFAIL("Regexp *+ operand could be empty");
 #endif
+
+        /* There's no point in trying to match something 0 length more than
+         * once except for extra side effects, which we don't have here since
+         * not POSTPONED */
```
Note that this could _replace_ the die/warn cases above. I feel it should at least replace the die.

khwilliamson replied:
@demerphq made this die for \K shortly before I was working on this code.  My sense was that he felt pretty strongly that no one should write code that had a >1 quantifier in \K.   I figure he knew better than I, and that I should change the existing behavior of the other cases, or that's my recollection.    Those actions are taken only for high maxes      `if (max > REG_INFTY/3) {` whereas there still is no point in doing it more than once even if we don't make a stink about it.  I left in the warning at that value to avoid changing any behavior.

hvds replied:
@demerphq never justified that despite requests; I feel pretty strongly that it should _not_ die.

khwilliamson replied:
I agree that it's better not to die

khwilliamson replied:
But when I tried it, I found that other things need to change.  I think then that this is outside the scope of this PR, so I am deferring it.

hvds commented on regcomp.c:
```diff
@@ -12763,32 +12763,35 @@ S_regpiece(pTHX_ RExC_state_t *pRExC_state, I32 *flagp, U32 depth)
-        if (min == 0 && max == REG_INFTY) {
+        if (max == REG_INFTY) {
+            if (min == 1) {
+                reginsert(pRExC_state, PLUS, ret, depth+1);
+                MARK_NAUGHTY(3);
+                goto done_main_op;
```
By moving the actual common bits further away from each other, I think this ends up making the commonality _less_ clear. I think it should at least retain the original order of testing `min == 0` first; even then, I'm not at all convinced this improves anything.

What really obscures the commonality is the 17-line chunk for the WILDCARD check. The real benefit to clarity would come from making that a terse check-and-call to a separate function where all the verbose commentary can live; but whether that's worth doing is also debatable.

khwilliamson replied:
I like the grouping of the * and + cases under a common upper-infinity one to show that commonality.  And that chunk of code I have moved to the end of the function and it is goto'd.  Since it croaks and doesn't return, that's sufficient.  I did reorder the * to be beore the +.  That makes the parallel code readily apparent.

hvds commented on regcomp.c:
```diff
@@ -12828,13 +12831,15 @@ S_regpiece(pTHX_ RExC_state_t *pRExC_state, I32 *flagp, U32 depth)
         MARK_NAUGHTY_EXP(1, 4);     /* compound interest */
     }
 
+    /* Finish up the {m,n} case */
```
This is rather the CURLY/CURLYX cases, since it also handles `?` and non-SIMPLE `+` and `*`.

khwilliamson replied:
ok

hvds reviewed (commented):
I've commented on a few minor things, but I don't see any correctness problems.

I'd propose also something like the below, which I feel improves clarity further; however under `/x` this change would allow whitespace between the `{` and the first digit - which I feel would be desirable, but only if we simultaneously allowed it elsewhere in `{m,n?}`.

I think it would also be a good idea to have a panic() fallback if we don't match any of `?*+{`, again even if only for the benefit of the reader.

```
--- a/regcomp.c
+++ b/regcomp.c
@@ -12633,9 +12633,20 @@ S_regpiece(pTHX_ RExC_state_t *pRExC_state, I32 *flagp, U32 depth)
     parse_start = RExC_parse;
 #endif
 
-    if (op == '{') {
+    nextchar(pRExC_state);
+
+    if (op == '*') {
+        min = 0;
+    }
+    else if (op == '+') {
+        min = 1;
+    }
+    else if (op == '?') {
+        min = 0; max = 1;
+    }
+    else if (op == '{') {
        maxpos = NULL;
-       next = RExC_parse + 1;
+       next = RExC_parse;
        while (isDIGIT(*next) || *next == ',') {
            if (*next == ',') {
                if (maxpos)
@@ -12651,7 +12662,6 @@ S_regpiece(pTHX_ RExC_state_t *pRExC_state, I32 *flagp, U32 depth)
         const char* endptr;
         if (!maxpos)
             maxpos = next;
-        RExC_parse++;
         if (isDIGIT(*RExC_parse)) {
             endptr = RExC_end;
             if (!grok_atoUV(RExC_parse, &uv, &endptr))
@@ -12693,20 +12703,6 @@ S_regpiece(pTHX_ RExC_state_t *pRExC_state, I32 *flagp, U32 depth)
                        *RExC_parse);
         }
     }
-    else {
-
-        nextchar(pRExC_state);
-
-        if (op == '*') {
-            min = 0;
-        }
-        else if (op == '+') {
-            min = 1;
-        }
-        else if (op == '?') {
-            min = 0; max = 1;
-        }
-    }
 
     if (! (flags & (HASWIDTH|POSTPONED))) {
         if (max > REG_INFTY/3) {
```

(khw) I don't know how to repsonde to this without editing it.I haven't fully studied this, but my first reaction was that I've been thinking about "Do we allow white space inside a \foo{...} in appropriate places, even without a /x.  I don't know.  We do allow it now in side \p{, but that is because that's the syntax that Unicode specifies, and I thought we better follow it to the letter.  So this is inconsistent with other braced expressions, but no one has ever complained to my knowledge, and maybe we should make those expressions consistent with \p{}.  Is there harm in doing so, except that it is a change, non-breaking AVAIK.  I think this release we can change {m,n} to allow a missing m, . and we could;d write it either way.

hvds commented:
(khw) I don't know how to repsonde to this without editing it.

You should be able to simply `quote reply`.

(khw) I haven't fully studied this, but my first reaction was that I've been thinking about "Do we allow white space inside a \foo{...} in appropriate places, even without a /x. I don't know.

I think `\foo{...}` (assuming `\foo` is a construct understood as special by the regexp engine) is different from `foo{...}` for a simple min/max range. I have always found it a wierd wart that whitespace is not allowed in the latter.

(khw) We do allow it now in side \p{, but that is because that's the syntax that Unicode specifies, and I thought we better follow it to the letter.  So this is inconsistent with other braced expressions, but no one has ever complained to my knowledge, and maybe we should make those expressions consistent with \p{}.  Is there harm in doing so, except that it is a change, non-breaking AVAIK.  I think this release we can change {m,n} to allow a missing m, . and we could;d write it either way.

I see no great benefit (and some small loss) in allowing a missing m. That seems decidedly orthogonal to whitespace.

lightsey commented on regcomp.c:
```diff
@@ -12637,192 +12645,201 @@ S_regpiece(pTHX_ RExC_state_t *pRExC_state, I32 *flagp, U32 depth)
-	    }
-	    else {
-		const regnode_offset w = reg_node(pRExC_state, WHILEM);
+        const char* endptr;
```
This generates a compiler error with clang 11:

```regcomp.c:12651:21: error: ISO C90 forbids mixing declarations and code [-Werror,-Wdeclaration-after-statement]```

khwilliamson replied:
fixed

khwilliamson committed:
regcomp.c: regpiece(): Convert to a switch() stmt

This makes the code easier to understand, I think.

khwilliamson committed:
regcomp.c: White-space only

Change indentation to correspond with new blocks formed by the previous
commit

khwilliamson committed:
regcomp.c: regpiece() Don't match 0 length more than once

There's no point in matching something more than once that doesn't
advance the parse, as long as there are no side effects beyond those
from the first match.  Those could only happen in code blocks, which
this code doesn't have.

khwilliamson committed:
regcomp.c: regpiece(): Pass along that is a code block

regpiece was failing to pass up to its caller that a piece of code
contained a code block, if that code was quantified.

I don't know what the consequences of this are

khwilliamson committed:
regcomp.c: regpiece: Move chunk of code for clarity

This changes an error branch to be goto'd out of the mainline code.  The
large chunk being in the middle obscured the comonality of the slightly
different non-error cases.

The branch is moved to the bottom of the routine, and croaks, so there
is no return.

This is a modification to a suggestion by Hugo van der Sanden.

khwilliamson committed:
regcomp.c: regpiece: swap order of conditionals

Its a bit more clearer to test the 0 case before the 1 case, and by
doing so it becomes visually easier to compare and contrast the the two
cases.
